### PR TITLE
Adding doc/dig_methods.rdoc and links to it

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Adds a full discussion of #dig, along with links from Array, Hash, Struct, and OpenStruct.

CSV::Table and CSV::Row are over in ruby/csv.  I'll get to them soon.

The art to the thing is to figure out how much (or how little) to say at each #dig.

### CODE REVIEW

File: hash.c [hunk 1]
```diff
@@ -5080,53 +5080,26 @@ rb_hash_any_p(int argc, VALUE *argv, VALUE hash)
...
- *    h = {foo: {bar: [:a, :b, :c]}}
- *    h.dig(:foo, :bar, 2) # => :c
+ *    h.dig(:foo, :bar, :BAZ) # => nil
  *
  *  ---
  *
- *  Raises an exception if any given key is invalid
+ *  Raises an exception if +key+ is invalid
  *  (see {Invalid Hash Keys}[#class-Hash-label-Invalid+Hash+Keys]):
  *    # Raises NoMethodError (undefined method `hash' for #<BasicObject>)
```
Comment (jeremyevans) [inline hash.c:5101]: If we are modifying documentation for this method, we should remove this example as well.

File: doc/dig_methods.rdoc [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  # Raises TypeError (Integer does not have #dig method):
+  h.dig(:foo, :bar)
+
+== What Else?
+
+The structure above has \Hash objects and \Array objects,
+both of which have instance method +dig+.
+
+Altogether there are six built-in Ruby classes that have method +dig+:
+- Array#dig: the first argument is an \Integer index.
```
Comment (jeremyevans) [inline doc/dig_methods.rdoc:72]: Only Array, Hash, and Struct are built-in (core classes).  The other three are stdlib classes.  I would change this to "three core classes and three classes in the standard library ".

File: hash.c (hunk 1 above)
Comment (BurdetteLamar) [reply hash.c]: Done.

File: doc/dig_methods.rdoc (hunk 2 above)
Comment (BurdetteLamar) [reply doc/dig_methods.rdoc:75]: I've elaborated this just a bit, to clarify which are where.

**Pull request merged**